pkg/kubernetes/watcher.go

```diff
@@ -0,0 +1,213 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+	"time"
+
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+
+	collector_metric "github.com/sustainable-computing-io/kepler/pkg/collector/metric"
+	"github.com/sustainable-computing-io/kepler/pkg/config"
+)
+
+const (
+	informerTimeout = time.Minute
+	podResourceType = "pods"
+)
+
+var (
+	regexReplaceContainerIDPrefix = regexp.MustCompile(`.*//`)
+	IsWatcherEnabled              = false
+	managedPods                   = make(map[string]bool)
+)
+
+type ObjListWatcher struct {
+	// Lock to syncronize the collector update with the watcher
+	Mx *sync.Mutex
+
+	k8sCli       *kubernetes.Clientset
+	ResourceKind string
+	informer     cache.SharedInformer
+	stopChannel  chan struct{}
+
+	// ContainersMetrics holds all container energy and resource usage metrics
+	ContainersMetrics *map[string]*collector_metric.ContainerMetrics
+}
+
+func newK8sClient() *kubernetes.Clientset {
+	var restConf *rest.Config
+	var err error
+	if config.KubeConfig == "" {
+		// creates the in-cluster config
+		restConf, err = rest.InClusterConfig()
+		klog.Infoln("Using in cluster k8s config")
+	} else {
+		// use the current context in kubeconfig
+		restConf, err = clientcmd.BuildConfigFromFlags("", config.KubeConfig)
+		klog.Infoln("Using out cluster k8s config: ", config.KubeConfig)
+	}
+	if err != nil {
+		klog.Infoln("%v", err)
+		return nil
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(restConf)
+	if err != nil {
+		klog.Fatalf("%v", err)
+	}
+	return clientset
+}
+
+func NewObjListWatcher() *ObjListWatcher {
+	w := &ObjListWatcher{
+		stopChannel:  make(chan struct{}),
+		k8sCli:       newK8sClient(),
+		ResourceKind: podResourceType,
+	}
+	if w.k8sCli == nil {
+		return w
+	}
+
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fmt.Sprintf("spec.nodeName=%s", collector_metric.NodeName) // to filter events per node
+	}
+	objListWatcher := cache.NewFilteredListWatchFromClient(
+		w.k8sCli.CoreV1().RESTClient(),
+		w.ResourceKind,
+		metav1.NamespaceAll,
+		optionsModifier,
+	)
+
+	w.informer = cache.NewSharedInformer(objListWatcher, nil, 0)
+	w.stopChannel = make(chan struct{})
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			w.handleUpdate(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.handleDeleted(obj)
+		},
+	})
+	IsWatcherEnabled = true
+	return w
+}
+
+func (w *ObjListWatcher) Run() {
+	if !IsWatcherEnabled {
+		return
+	}
+	go w.informer.Run(w.stopChannel)
```

Comment:
this implementation is referenced from official sdk, exits in safety?

Reply:
Sorry, can you explain more about this?

Reply:
I want to ensure the go routine safety here.